File: Ningin/src/math/vector3.cpp
#include "Vector3.h"
#include "Math.h"
#include "Vector2.h"

// Static Constants Initialization
const Vector3 Vector3::UNIT_X = Vector3(1.0f, 0.0f, 0.0f);
const Vector3 Vector3::UNIT_Y = Vector3(0.0f, 1.0f, 0.0f);
const Vector3 Vector3::UNIT_Z = Vector3(0.0f, 0.0f, 1.0f);
const Vector3 Vector3::UNIT_XY = Vector3(1.0f, 1.0f, 0.0f);
const Vector3 Vector3::UNIT_XZ = Vector3(1.0f, 0.0f, 1.0f);
const Vector3 Vector3::UNIT_YZ = Vector3(0.0f, 1.0f, 1.0f);
const Vector3 Vector3::ZERO = Vector3(0.0f, 0.0f, 0.0f);
const Vector3 Vector3::ONE = Vector3(1.0f, 1.0f, 1.0f);

Vector3::Vector3(float x, float y, float z) : x(x), y(y), z(z) {}

Vector3::Vector3(Vector2& vec) : x(vec.x), y(vec.y), z(0.0f) {}

void Vector3Abs(Vector3& vec, Vector3& out)
{
	out.x = std::abs(vec.x);
	out.y = std::abs(vec.y);
	out.z = std::abs(vec.z);
}

void Vector3SquareRoot(Vector3& vec, Vector3& out)
{
	out.x = sqrt(vec.x);
	out.y = sqrt(vec.y);
	out.z = sqrt(vec.z);
}

void Vector3Add(Vector3& v1, Vector3& v2, Vector3& out)
{
	out.x = v1.x + v2.x;
	out.y = v1.y + v2.y;
	out.z = v1.z + v2.z;
}

float Vector3Magnitude(Vector3& vec)
{
	return sqrt(Vector3Dot(vec, vec));
}

void Vector3Normalize(Vector3& vec, Vector3& out)
{
	float length = Vector3Magnitude(vec);
	out = Vector3(vec.x / length, vec.y / length, vec.z/length);
}

float Vector3Dot(Vector3& v1, Vector3& v2)
{
	return v1.x * v2.x + v1.y * v2.y + v1.z * v2.z;
}

float Vector3Distance(Vector3& v1, Vector3& v2)
{
	Vector3 vec = Vector3(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z);
	return Vector3Magnitude(vec);
}

float Vector3DistanceSquared(Vector3& v1, Vector3& v2)
{
	float distance = Vector3Distance(v1, v2);
	return distance * distance;
}

float Vector3MagnitudeSquared(Vector3& vec)
{
	return Vector3Dot(vec, vec);
}

void Vector3Max(Vector3& v1, Vector3& v2, Vector3& out)
{
	out = Vector3(max(v1.x, v2.x), max(v1.y, v2.y), max(v1.z, v2.z));
}

void Vector3Min(Vector3& v1, Vector3& v2, Vector3& out)
{
	out = Vector3(min(v1.x, v2.x), min(v1.y, v2.y), min(v1.z, v2.z));
}

void Vector3Clamp(Vector3& vecToClamp, Vector3& min, Vector3& max, Vector3& out)
{
	out.x = Clamp(vecToClamp.x, min.x, max.x);
	out.y = Clamp(vecToClamp.y, min.y, max.y);
	out.z = Clamp(vecToClamp.z, min.z, max.z);
}

void Vector3Negate(Vector3& vec, Vector3& out) {
	out = Vector3(-vec.x, -vec.y, -vec.z);
}

void Vector3Subtract(Vector3& v1, Vector3& v2, Vector3& out){
	out = Vector3(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z);
}

void Vector3MultiplyByFloat(Vector3& vec, float a, Vector3& out){
	out = Vector3(vec.x * a, vec.y * a, vec.z * a);
}

void Vector3Multiply(Vector3& v1, Vector3& v2, Vector3& out){
	out = Vector3(v1.x * v2.x, v1.y * v2.y, v1.z * v2.z);
}

void Vector3Divide(Vector3& v1, Vector3& v2, Vector3& out){
	out = Vector3(v1.x / v2.x, v1.y / v2.y, v1.z / v2.z);
}

void Vector3DivideByFloat(Vector3& vec, float a, Vector3& out){
	out = Vector3(vec.x / a, vec.y / a, vec.z / a);

}

bool Vector3Equals(Vector3& v1, Vector3& v2){
	return v1.x == v2.x && v1.y == v2.y && v1.z == v2.z;
}

bool Vector3NotEqual(Vector3& v1, Vector3& v2){
	return !(Vector3Equals(v1, v2));
}

bool Vector3L(Vector3& v1, Vector3& v2){
	return v1.x < v2.x && v1.y < v2.y && v1.z < v2.z;
}

bool Vector3LE(Vector3& v1, Vector3& v2){
	return v1.x <= v2.x && v1.y <= v2.y && v1.z <= v2.z;

}

bool Vector3G(Vector3& v1, Vector3& v2){
	return v1.x > v2.x && v1.y > v2.y && v1.z > v2.z;
}

bool Vector3GE(Vector3& v1, Vector3& v2){
	return v1.x >= v2.x && v1.y >= v2.y && v1.z >= v2.z;
}

void Vector3ToString(Vector3& vec, string& out)
{
	out = format("Vector2 - x: {}, y: {}, z: {}", vec.x, vec.y, vec.z);
}

void Vector3FromVector2(Vector2& vec, Vector3& out)
{
	out = Vector3(vec.x, vec.y);
}

File: NinginCore/math/Vector3.cs
namespace NinginCore
{
    public struct Vector3
    {
        public static Vector3 UNIT_X = new Vector3(1.0f, 0.0f, 0.0f);
        public static Vector3 UNIT_Y = new Vector3(0.0f, 1.0f, 0.0f);
        public static Vector3 UNIT_Z = new Vector3(0.0f, 0.0f, 1.0f);
        public static Vector3 ZERO = new Vector3(0.0f, 0.0f, 0.0f);
        public static Vector3 ONE = new Vector3(1.0f, 1.0f, 1.0f);

        public float x, y, z;

        public Vector3(float x = 0.0f, float y = 0.0f, float z = 0.0f)
        {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public Vector3 Abs()
        {
            InternalCalls.Vector3Abs(ref this, out Vector3 result);
            return result;
        }

        public Vector3 Negate()
        {
            InternalCalls.Vector3Negate(ref this, out Vector3 result);
            return result;
        }
        public Vector3 Scale(float scaler)
        {
            InternalCalls.Vector3MultiplyByFloat(ref this, scaler, out Vector3 result);
            return result;
        }

        public Vector3 Sqrt()
        {
            InternalCalls.Vector3SquareRoot(ref this, out Vector3 result);
            return result;
        }

        public Vector3 Add(Vector3 Other)
        {
            InternalCalls.Vector3Add(ref this, ref Other, out Vector3 result);
            return result;
        }

        public Vector3 Clamp(Vector3 min, Vector3 max)
        {
            InternalCalls.Vector3Clamp(ref this, ref min, ref max, out Vector3 result);
            return result;
        }

        public Vector3 Normalize()
        {
            InternalCalls.Vector3Normalize(ref this, out Vector3 result);
            return result;
        }
        public Vector3 Inverse()
        {
            InternalCalls.Vector3Inverse(ref this, out Vector3 result);
            return result;
        }

        public Vector3 CrossProduct(Vector3 rightVector)
        {
            InternalCalls.Vector3Cross(ref this, ref rightVector, out Vector3 result);
            return result;
        }

        public float Distance(Vector3 Other) => InternalCalls.Vector3Distance(ref this, ref Other);
        public float Dot(Vector3 Other) => InternalCalls.Vector3Dot(ref this, ref Other);
        public float Magnitude() => InternalCalls.Vector3Magnitude(ref this);
        public float MagnitudeSquared() => InternalCalls.Vector3MagnitudeSquared(ref this);
        public float DistanceSquared(Vector3 Other)
            => InternalCalls.Vector3DistanceSquared(ref this, ref Other);
        public override string ToString()
        {
            InternalCalls.MonoVector3ToString(ref this, out string result);
            return result;
        }

        public static float Dot(Vector3 v1, Vector3 v2) => InternalCalls.Vector3Dot(ref v1, ref v2);
        public static float Magnitude(Vector3 vec) => InternalCalls.Vector3Magnitude(ref vec);
        public static float MagnitudeSquared(Vector3 vec) => InternalCalls.Vector3MagnitudeSquared(ref vec);

        public static float Distance(Vector3 v1, Vector3 v2)
            => InternalCalls.Vector3Distance(ref v1, ref v2);

        public static float DistanceSquared(Vector3 v1, Vector3 v2)
            => InternalCalls.Vector3DistanceSquared(ref v1, ref v2);

        public static Vector3 Max(Vector3 V1, Vector3 V2)
        {
            InternalCalls.Vector3Max(ref V1, ref V2, out Vector3 result);
            return result;
        }

        public static Vector3 Min(Vector3 V1, Vector3 V2)
        {
            InternalCalls.Vector3Min(ref V1, ref V2, out Vector3 result);
            return result;
        }

        public static Vector3 Clamp(Vector3 to_clamp, Vector3 min, Vector3 max)
        {
            InternalCalls.Vector3Clamp(ref to_clamp, ref min, ref max, out Vector3 result);
            return result;
        }

        public static Vector3 Normalize(Vector3 vec)
        {
            InternalCalls.Vector3Normalize(ref vec, out Vector3 result);
            return result;
        }

        public static Vector3 Inverse(Vector3 vec)
        {
            InternalCalls.Vector3Inverse(ref vec, out Vector3 result);
            return result;
        }

        // Operators
        public static Vector3 operator -(Vector3 vec)
        {
            InternalCalls.Vector3Negate(ref vec, out Vector3 result);
            return result;
        }

        public static Vector3 operator +(Vector3 v1, Vector3 v2)
        {
            InternalCalls.Vector3Add(ref v1, ref v2, out Vector3 result);
            return result;
        }

        public static Vector3 operator -(Vector3 v1, Vector3 v2)
        {
            InternalCalls.Vector3Subtract(ref v1, ref v2, out Vector3 result);
            return result;
        }

        public static Vector3 operator *(Vector3 v1, float a)
        {
            InternalCalls.Vector3MultiplyByFloat(ref v1, a, out Vector3 result);
            return result;
        }

        public static Vector3 operator *(float a, Vector3 v1)
        {
            InternalCalls.Vector3MultiplyByFloat(ref v1, a, out Vector3 result);
            return result;
        }

        public static Vector3 operator *(Vector3 v1, Vector3 v2)
        {
            InternalCalls.Vector3Multiply(ref v1, ref v2, out Vector3 result);
            return result;
        }

        public static Vector3 operator /(Vector3 v1, Vector3 v2)
        {
            InternalCalls.Vector3Divide(ref v1, ref v2, out Vector3 result);
            return result;
        }

        public static Vector3 operator /(Vector3 vec, float a)
        {
            InternalCalls.Vector3DivideByFloat(ref vec, a, out Vector3 result);
            return result;
        }

        public static Vector3 operator /(float a, Vector3 vec)
        {
            InternalCalls.Vector3Inverse(ref vec, out Vector3 inversedVec);
            InternalCalls.Vector3DivideByFloat(ref inversedVec, 1 / a, out Vector3 result);
            return result;
        }

        public static bool operator <(Vector3 v1, Vector3 v2) => InternalCalls.Vector3L(ref v1, ref v2);
        public static bool operator <=(Vector3 v1, Vector3 v2) => InternalCalls.Vector3LE(ref v1, ref v2);
        public static bool operator >(Vector3 v1, Vector3 v2) => InternalCalls.Vector3G(ref v1, ref v2);
        public static bool operator >=(Vector3 v1, Vector3 v2) => InternalCalls.Vector3GE(ref v1, ref v2);

        public static bool operator ==(Vector3 v1, Vector3 v2)
            => InternalCalls.Vector3Equals(ref v1, ref v2);

        public static bool operator !=(Vector3 v1, Vector3 v2)
            => InternalCalls.Vector3NotEqual(ref v1, ref v2);

        public static Vector3 FromVector3(Vector2 vec)
        {
            InternalCalls.Vector3FromVector2(ref vec, out Vector3 result);
            return result;
        }

        public override bool Equals(object obj)
        {
            if (obj is Vector3 vector)
            {
                return this == vector;
            }

            return false;
        }

        public override int GetHashCode() => new { x, y, z }.GetHashCode();
    }
}
